pkg/chain/ethlike/config.go

```diff
@@ -0,0 +1,74 @@
+package ethlike
+
+import "math/big"
+
+// Account is a struct that contains the configuration for accessing an
+// ETH-like network and a contract on the network.
+type Account struct {
+	// Keyfile is a full path to a key file.  Normally this file is one of the
+	// imported keys in your local chain node.  It can normally be found in
+	// a directory <some-path>/data/keystore/ and starts with its creation date
+	// "UTC--.*".
+	KeyFile string
+
+	// KeyFilePassword is the password used to unlock the account specified in
+	// KeyFile.
+	KeyFilePassword string
+}
+
+// CommonConfig is a struct that contains the configuration needed to connect
+// to an ETH-like node. This information will give access to an ETH-like network.
+type CommonConfig struct {
+	// Example: "ws://192.168.0.157:8546".
+	URL string
+
+	// Example: "http://192.168.0.157:8545".
+	URLRPC string
+
+	// A  map from contract names to contract addresses.
+	ContractAddresses map[string]string
+
+	Account Account
+
+	// MiningCheckInterval is the interval in which transaction
+	// mining status is checked. If the transaction is not mined within this
+	// time, the gas price is increased and transaction is resubmitted.
+	MiningCheckInterval int
+
+	// MaxGasPrice specifies the maximum gas price the client is
+	// willing to pay for the transaction to be mined. The offered transaction
+	// gas price can not be higher than the max gas price value. If the maximum
+	// allowed gas price is reached, no further resubmission attempts are
+	// performed.
+	MaxGasPrice *Value
+
+	// RequestsPerSecondLimit sets the maximum average number of requests
+	// per second which can be executed against the ETH-like node.
+	// All types of chain requests are rate-limited,
+	// including view function calls.
+	RequestsPerSecondLimit int
+
+	// ConcurrencyLimit sets the maximum number of concurrent requests which
+	// can be executed against the ETH-like node at the same time.
+	// This limit affects all types of chain requests,
+	// including view function calls.
+	ConcurrencyLimit int
+
+	// BalanceAlertThreshold defines a minimum value of the operator's
+	// account balance below which an alert will be triggered.
+	BalanceAlertThreshold *Value
+}
+
+// BalanceAlertThresholdValue returns the `BalanceAlertThreshold` integer value.
+func (cg *CommonConfig) BalanceAlertThresholdValue() *big.Int {
+	if cg.BalanceAlertThreshold != nil {
+		return cg.BalanceAlertThreshold.Int
+	}
+
+	return nil
+}
```

Comment:
There is an interface introduced by https://github.com/keep-network/keep-ecdsa/pull/686 which makes use of that method. This is the reasoning behind that change.

Reply:
For this function, I'd consider moving `balance_monitor.go` to `ethlike` of `keep-commons` and have `celo` and `ethereum` version accepting chain-specific type, as noted [here](https://github.com/keep-network/keep-common/pull/66#discussion_r580855694).

Reply:
But that doesn't change anything in terms of `chainConfig` usage as we still need to run the balance monitor (`initializeBalanceMonitoring`). If we want to avoid duplications of that code, we should leave it in the `start.go` command which is now chain-agnostic. So, we still need a way to abstract the specific chain config.

Reply:
I am not sure if I follow correctly but let me approach it from another side.

If we leave `BalanceAlertThresholdValue *big.Int` we are leaking abstractions. `celo.Wei` and `ethereum.Wei` are 1e18 precision `ethlike.Token` but this is not something `keep-ecdsa/cmd/start.go` should be aware of. Inside Celo or Ethereum balance monitor, assuming this monitor lives in the same package as `celo.Wei` or `ethereum.Wei`, we can do more and do some package-internal assumptions.

We can have a "parent" `balance_monitor.go` in `ethlike` package that will work with the appropriate chain client interface and having that interface guarantee the returned balance `*big.Int` is 1e18 precision token. We can have `celo` and `ethereum` balance monitors inheriting from `ethlike`'s one and accepting `celo.Wei` or `ethereum.Wei` as the minimum balance parameter. Internally, they can do a conversion to 1e18 `*big.Int` to have it working with `ethlike` balance monitor. 

The glue code will probably have to live in `connector_*.go` and `start.go` will not even touch the initialization but this will probably be the case anyway given that we need to set a `defaultBalanceAlertThreshold` that will be different for Celo and Ethereum.

Reply:
First, both `celo.Wei` and `ethereum.Wei` are just utilities to parse values from the config file. They are not intended to represent chain tokens everywhere. Second, AFAIK all our code which deals with token values assumes those values are represented in the chain default unit (`wei` in case of Ethereum and Celo). Now if we look at the config and invoke `BalanceAlertThresholdValue` we just receive that value in the default chain unit. Nothing leaks at this point. 

Now, this is the interface of a balance monitor:
```
type BalanceMonitor interface {
	// Observe starts a process which checks the address balance with the given
	// tick and triggers an alert in case the balance falls below the
	// alert threshold value.
	Observe(
		ctx context.Context,
		address string,
		alertThreshold *big.Int,
		tick time.Duration,
	)
}
```
It is completely independent of the `ethlike.Token`. It just expects a `big.Int` alert threshold. Following the assumption that the threshold is represented in the default unit, everything is fine. Both implementations also take the current balance as `*big.Int` and assume it is represented in the default unit.

Regarding everything I wrote above, I don't understand why we may want to complicate things and introduce `ethlike.Token` to the balance monitor. If we want to do this we should also find all places where we operate on token values and do the same.

Reply:
Correct, everything you say is right. I do not think we should eventually represent price/balance as `*big.Int`. This was ~fine when Ethereum was the only host chain. Now, as we are moving forward doing this refactor and the refactor of chain interface, I do think price/balance should be represented by a value object. Internally, this value object can use `*big.Int` for ETH-like chains but may as well use something else for non-ETH-like chains.

That's the same situation as with `common.Address` in chain interface - we would be able to pass it around as `string` but it is not as elegant as a value object and does not allow for elegant conversions. For price/balance, we may, for example, have a need of calling some contract passing `Gwei` value, or have a need to convert from CELO to cUSD. If we pass around `*big.Int` we lose the information about currency and denomination. If we pass price/balance as a value object, we have all of it available. So yes, we should find all other places. Eventually, not right now.

Short-term, I'd try to keep `*big.Int` inside `ethlike`/`ethereum/`celo` or at least, try not to reduce CELO wei and Ethereum wei into a `*big.Int` in a code outside of the chain implementation what `chainConfig` interface in `keep-ecdsa` currently does.

IMO, `balance_monitor.go` should live in `keep-common` anyway. Right now, Ethereum balance monitor is in `keep-core` and Celo balance monitor is in `keep-ecdsa`. If we do it, there is a short path from having Celo and Ethereum versions of balance monitor accept `celo.Wei` or `ethereum.Wei`, especially that we split them in this PR.

Reply:
> IMO, balance_monitor.go should live in keep-common anyway. Right now, Ethereum balance monitor is in keep-core and Celo balance monitor is in keep-ecdsa. If we do it, there is a short path from having Celo and Ethereum versions of balance monitor accept celo.Wei or ethereum.Wei, especially that we split them in this PR.

Done in https://github.com/keep-network/keep-common/commit/e715d7f415d40236121dd71ac20d8e4e33cf3c0b. However, I'm not sure if we will be able to accept `celo.Wei` and `ethereum.Wei` separately in an easy way as both balance monitors need to implement a common `BalanceMonitor` interface.

Reply:
If we are talking about `BalanceMonitor` interface in `keep-core/pkg/chain/chain.go`, it seems that we reached the same conclusion here and in [unchainded refactor](https://github.com/keep-network/keep-ecdsa/pull/693#discussion_r576783818) PR. This component may be Ethereum (or Ethereum-like) specific concept so I think we can get rid of it from `Chain` interface and have it called in connectors with a specific currency passed to `Observe` function.

Reply:
I refactored the balance monitors according to our discussion.